src/core/dbus-service.c

```diff
@@ -718,6 +748,56 @@ static int bus_service_set_transient_property(
         if (streq(name, "ReloadSignal"))
                 return bus_set_transient_reload_signal(u, name, &s->reload_signal, message, flags, error);
 
+        if (streq(name, "ExtraFileDescriptor")) {
+                int fd;
+                const char *fdname;
+
+                r = sd_bus_message_enter_container(message, 'a', "(hs)");
+                if (r < 0)
+                        return r;
+
+                while ((r = sd_bus_message_read(message, "(hs)", &fd, &fdname)) > 0) {
+                        _cleanup_(file_descriptor_freep) FileDescriptor *extra_fd = NULL;
+                        _cleanup_close_ int dup_fd = -EBADF;
+
+                        extra_fd = new(FileDescriptor, 1);
+                        if (!extra_fd)
+                                return -ENOMEM;
+
+                        dup_fd = fcntl(fd, F_DUPFD_CLOEXEC, 3);
+                        if (dup_fd < 0)
+                                return -errno;
+
+                        *extra_fd = (FileDescriptor) {
+                                .fd = dup_fd,
+                                .fdname = strdup(fdname),
+                        };
+
+                        if (!extra_fd->fdname)
+                                return -ENOMEM;
+
+                        r = file_descriptor_validate(extra_fd);
+                        if (r < 0)
+                                return r;
+
+                        if (UNIT_WRITE_FLAGS_NOOP(flags))
+                                continue;
+
+                        // Reset dup_fd to EBADF to ensure we don't close FD
+                        TAKE_FD(dup_fd);
```

Comment:
So you can move this into the initialization of `extra_fd` and then let its destructor take care of cleaning up or not.